### PR TITLE
Squirrel bridge v3

### PR DESCRIPTION
**Pull request by emma-miler**
Adds `call` and `schedule_call` with completely automatic variadic argument pushing and threadsafe asynchronous calls
Depends on https://github.com/R2Northstar/NorthstarMods/pull/525

### CODE REVIEW

File: NorthstarDLL/squirrel.h [hunk 1]
```diff
@@ -1,140 +1,19 @@
...
+
+// This base class means that only the templated functions have to be rebuilt for each template instance
+// Cuts down on compile time by ~5 seconds
+class SquirrelManagerBase
 {
-  private:
+  public:
 	std::vector<SQFuncRegistration*> m_funcRegistrations;
 
   public:
```
Comment (Mauler125) [inline NorthstarDLL/squirrel.h:19]: Extraneous access modifier.

File: NorthstarDLL/squirrel.h [hunk 2]
```diff
@@ -340,4 +221,203 @@ template <ScriptContext context> class SquirrelManager
...
+		// Calling this function while the VM is running is likely to result in a crash due to stack destruction
+		// If you want to call into squirrel asynchronously, use `schedule_call` instead
+
+		if (!m_pSQVM || !m_pSQVM->sqvm)
+		{
+			spdlog::error(
+				"{} was called on context {} while VM was not initialized. This will crash", __FUNCTION__, GetContextName(context));
+		}
+
+		SQObject* functionobj = new SQObject();
```
Comment (Mauler125) [inline NorthstarDLL/squirrel.h:264]: You allocate new memory, but should this be freed?
If it should, does the internal engine function 'pushSQObject' free this memory?

If it shouldn't, add a comment.

File: NorthstarDLL/squirrel.h [hunk 3]
```diff
@@ -340,4 +221,203 @@ template <ScriptContext context> class SquirrelManager
...
+		// Warning!
+		// This function assumes the squirrel VM is stopped/blocked at the moment of call
+		// Calling this function while the VM is running is likely to result in a crash due to stack destruction
+		// If you want to call into squirrel asynchronously, use `schedule_call` instead
+		if (!m_pSQVM || !m_pSQVM->sqvm)
+		{
+			spdlog::error(
+				"{} was called on context {} while VM was not initialized. This will crash", __FUNCTION__, GetContextName(context));
+		}
+		SQObject* functionobj = new SQObject();
```
Comment (Mauler125) [inline NorthstarDLL/squirrel.h:286]: You allocate new memory, but should this be freed?
If it should, does the internal engine function 'pushSQObject' free this memory?

If it shouldn't, add a comment.

File: NorthstarDLL/squirrelclasstypes.h [hunk 4]
```diff
@@ -0,0 +1,239 @@
+#pragma once
+#include "pch.h"
+#include "squirreldatatypes.h"
+#include <map>
```
Comment (Mauler125) [inline NorthstarDLL/squirrelclasstypes.h:4]: Extraneous include; this is already included in the PCH file.

File: NorthstarDLL/clientchathooks.cpp [hunk 5]
```diff
@@ -16,29 +16,22 @@ void, __fastcall, (void* self, const char* message, int inboxId, bool isTeam, bo
...
+	const char* payload = message;
+	if (isCustom)
+	{
+		type = message[0];
+		payload = message + 1;
 	}
-	else
+
+	SQRESULT res =
+		g_pSquirrel<ScriptContext::CLIENT>->call("CHudChat_ProcessMessageStartThread", (int)senderId - 1, payload, isTeam, isDead, type);
```
Comment (Mauler125) [inline NorthstarDLL/clientchathooks.cpp:33]: If you go modern C++, use C++ style casts instead.

File: NorthstarDLL/serverchathooks.cpp [hunk 6]
```diff
@@ -37,31 +37,26 @@ void(__fastcall* MessageWriteBool)(bool bValue);
...
-	if (g_pSquirrel<ScriptContext::SERVER>->setupfunc("CServerGameDLL_ProcessMessageStartThread") != SQRESULT_ERROR)
-	{
-		g_pSquirrel<ScriptContext::SERVER>->pushinteger(g_pSquirrel<ScriptContext::SERVER>->m_pSQVM->sqvm, (int)nSenderPlayerIndex - 1);
-		g_pSquirrel<ScriptContext::SERVER>->pushstring(g_pSquirrel<ScriptContext::SERVER>->m_pSQVM->sqvm, text);
-		g_pSquirrel<ScriptContext::SERVER>->pushbool(g_pSquirrel<ScriptContext::SERVER>->m_pSQVM->sqvm, isTeam);
-		g_pSquirrel<ScriptContext::SERVER>->call(g_pSquirrel<ScriptContext::SERVER>->m_pSQVM->sqvm, 3);
-	}
-	else
-		_CServerGameDLL__OnReceivedSayTextMessage(self, nSenderPlayerIndex, text, isTeam);
+	g_pSquirrel<ScriptContext::SERVER>->call("CServerGameDLL_ProcessMessageStartThread", (int)senderPlayerId - 1, text, isTeam);
```
Comment (Mauler125) [inline NorthstarDLL/serverchathooks.cpp:56]: If you go modern C++, use C++ style casts instead.

**Comment by GeckoEidechse** [comment]
@emma-miler merge conflicts now ^^"

**Comment by emma-miler** [comment]
fek

**Comment by emma-miler** [comment]
@GeckoEidechse fixed

**Comment by emma-miler** [comment]
How to test:
Run this build and use the associated mods PR. Then simply run `script_ui NSTestMessageBuffer()` on the main menu. You should see some debug output. This means its working

**Comment by Erlite** [comment]
Will test with the HTTP PR in a few.

**Review by ASpoonPlaysGames — APPROVED** [review]
Works in testing

**Comment by emma-miler** [comment]
Due to a restructuring in the code, testing this directly is now no longer possible. Sorry.

File: NorthstarDLL/squirrel.h [hunk 7]
```diff
@@ -103,63 +31,23 @@ static constexpr int operator|(int first, ScriptContext second)
 	return first + (1 << static_cast<int>(second));
 }
 
+#include "squirrelclasstypes.h"
```
Comment (BobTheBob9) [inline NorthstarDLL/squirrel.h:34]: these should probably be at the top of the file

File: NorthstarDLL/squirrel.h [hunk 8]
```diff
@@ -103,63 +31,23 @@ static constexpr int operator|(int first, ScriptContext second)
...
-template <ScriptContext context> class SquirrelManager
+void schedule_call_external(ScriptContext context, const char* func_name, SquirrelMessage_External_Pop function);
+
+// This base class means that only the templated functions have to be rebuilt for each template instance
+// Cuts down on compile time by ~5 seconds
+class SquirrelManagerBase
 {
-  private:
+  public:
 	std::vector<SQFuncRegistration*> m_funcRegistrations;
```
Comment (BobTheBob9) [inline NorthstarDLL/squirrel.h:49]: should be private

File: NorthstarDLL/squirrel.cpp [hunk 9]
```diff
@@ -482,6 +508,41 @@ template <ScriptContext context> void StubUnsafeSQFuncs()
 	}
 }
 
+template <ScriptContext context> void SquirrelManager<context>::ProcessMessageBuffer()
+{
+	auto maybe_message = messageBuffer->pop();
```
Comment (BobTheBob9) [inline NorthstarDLL/squirrel.cpp:513]: we don't really use snake case

File: NorthstarDLL/squirrelclasstypes.h [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+
+typedef void (*SquirrelMessage_External_Pop)(HSquirrelVM* sqvm);
+typedef void (*sq_schedule_call_externalType)(ScriptContext context, const char* funcname, SquirrelMessage_External_Pop function);
+
+class SquirrelMessage
+{
+  public:
+	std::string function_name;
+	FunctionVector args;
+	bool is_external = false;
```
Comment (BobTheBob9) [inline NorthstarDLL/squirrelclasstypes.h:127]: snake case

File: NorthstarDLL/squirrelclasstypes.h [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+		}
+		else
+		{
+			return std::nullopt;
+		}
+	}
+
+	void unwind()
+	{
+		auto maybe_message = this->pop();
```
Comment (BobTheBob9) [inline NorthstarDLL/squirrelclasstypes.h:156]: snake case

File: NorthstarDLL/squirrelclasstypes.h [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+
+// sq stack userpointer funcs
+typedef void* (*sq_createuserdataType)(HSquirrelVM* sqvm, SQInteger iSize);
+typedef SQRESULT (*sq_setuserdatatypeidType)(HSquirrelVM* sqvm, SQInteger iStackpos, uint64_t iTypeId);
+
+// sq misc entity funcs
+typedef void* (*sq_getentityfrominstanceType)(CSquirrelVM* sqvm, SQObject* pInstance, char** ppEntityConstant);
+typedef char** (*sq_GetEntityConstantType)();
+
+typedef int (*sq_getSquirrelFunctionType)(HSquirrelVM* sqvm, const char* name, SQObject* returnObj, const char* signature);
```
Comment (BobTheBob9) [inline NorthstarDLL/squirrelclasstypes.h:234]: not massive on this name

File: NorthstarDLL/squirrel.h [hunk 13]
```diff
@@ -364,4 +255,205 @@ template <ScriptContext context> class SquirrelManager
...
+		// This function schedules a call to be executed on the next frame
+		// This is useful for things like threads and plugins, which do not run on the main thread
+		FunctionVector function_vector;
+		SqRecurseArgs<context>(function_vector, args...);
+		SquirrelMessage message = {funcname, function_vector};
+		messageBuffer->push(message);
+		return message;
+	}
+
+	SquirrelMessage schedule_call(std::string funcname)
```
Comment (BobTheBob9) [inline NorthstarDLL/squirrel.h:275]: snake case

File: NorthstarDLL/squirrel.h [hunk 14]
```diff
@@ -364,4 +255,205 @@ template <ScriptContext context> class SquirrelManager
...
+	{
+		// This function schedules a call to be executed on the next frame
+		// This is useful for things like threads and plugins, which do not run on the main thread
+		FunctionVector function_vector = {};
```
Comment (BobTheBob9) [inline NorthstarDLL/squirrel.h:279]: snake case

File: NorthstarDLL/squirrel.h [hunk 15]
```diff
@@ -364,4 +255,205 @@ template <ScriptContext context> class SquirrelManager
...
+		SQObject functionobj {};
+		int result = g_pSquirrel<context>->sq_getSquirrelFunction(g_pSquirrel<context>->m_pSQVM->sqvm, funcname, &functionobj, 0);
+		if (result != 0) // This func returns 0 on success for some reason
+		{
+			return SQRESULT_ERROR;
+		}
+		g_pSquirrel<context>->pushobject(g_pSquirrel<context>->m_pSQVM->sqvm, &functionobj); // Push the function object
+		g_pSquirrel<context>->pushroottable(g_pSquirrel<context>->m_pSQVM->sqvm); // Push root table
+
+		FunctionVector function_vector;
```
Comment (BobTheBob9) [inline NorthstarDLL/squirrel.h:329]: snake case

File: NorthstarDLL/serverchathooks.cpp [hunk 16]
```diff
@@ -37,31 +37,27 @@ void(__fastcall* MessageWriteBool)(bool bValue);
 bool bShouldCallSayTextHook = false;
 // clang-format off
 AUTOHOOK(_CServerGameDLL__OnReceivedSayTextMessage, server.dll + 0x1595C0,
-void, __fastcall, (CServerGameDLL* self, unsigned int nSenderPlayerIndex, const char* text, bool isTeam))
+void, __fastcall, (CServerGameDLL* self, unsigned int senderPlayerId, const char* text, bool isTeam))
```
Comment (BobTheBob9) [inline NorthstarDLL/serverchathooks.cpp:40]: why was this renamed? not necessary and arguably worse name than before

File: NorthstarDLL/squirrel.cpp [hunk 17]
```diff
@@ -138,6 +140,25 @@ const char* SQTypeNameFromID(int type)
 	return "";
 }
 
+// Allows for generating squirrelmessages from plugins.
+// Not used in this version, but will be used later
+void schedule_call_external(ScriptContext context, const char* func_name, SquirrelMessage_External_Pop function)
```
Comment (BobTheBob9) [inline NorthstarDLL/squirrel.cpp:145]: snake case

**Review by BobTheBob9 — CHANGES_REQUESTED** [review]
variable naming as a whole needs work and should follow how we name stuff in other places, aside from that looks fine

File: NorthstarDLL/serverchathooks.cpp (hunk 16 above)
Comment (emma-miler) [reply NorthstarDLL/serverchathooks.cpp:40]: previously thought this was Index, but turns out it's actually ID

**Review by BobTheBob9 — APPROVED** [review]
